Patient treatment phases

alintreznai opened the pull request:
This PR introduces treatment phases support for patients.

Commit:
Introducing patient stage model

sztamas commented on rdrf/registry/patients/migrations/0037_patient_stage_data_migration.py:
```diff
@@ -0,0 +1,31 @@
+from django.db import migrations, models
+
+
+def add_patient_stages(apps, schema_editor):
```
We don't set up data in data migrations in this project, we use "initial data" instead.
See: https://github.com/eresearchqut/trrf/tree/next_release/rdrf/rdrf/initial_data

alintreznai replied:
Updated.

sztamas commented on rdrf/registry/patients/models.py:
```diff
@@ -1118,6 +1123,11 @@ def family(self):
 
 
+class PatientStage(models.Model):
+    name = models.CharField(max_length=200, null=False)
```
`null=False` is the default. Can be omitted.

alintreznai replied:
Updated.

sztamas commented on rdrf/registry/patients/models.py:
```diff
@@ -1118,6 +1123,11 @@ def family(self):
 
+class PatientStage(models.Model):
+    name = models.CharField(max_length=200, null=False)
+    previous = models.ManyToManyField('PatientStage', related_name='from_stage')
```
Maybe use `'self'` instead of `'PatientStage'` to make it clear it is a self-reference.

The name/related name could be clearer? Maybe `allowed_prev_stages` and `allowed_next_stages`?

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/initial_data/DEV.py:
```diff
@@ -18,6 +18,7 @@
     'genes_smaller_sample',
     'sample_registry',
     'iprestrict_permissive',
+    'patient_stage'
```
Needed? `patient_stage` is part of `reference_data` now.

sztamas commented on rdrf/rdrf/initial_data/__init__.py:
```diff
@@ -1,5 +1,5 @@
 from . import (base, groups, users, sample_registry, sample_laboratories,
-               genes, genes_smaller_sample, DEV, iprestrict_permissive, reference_data)
+               genes, genes_smaller_sample, DEV, iprestrict_permissive, reference_data, patient_stage)
```
Needed? `patient_stage` is part of `reference_data` now.

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/initial_data/patient_stage.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def load_data(**kwargs):
+    informed_consent = create_stage('InformedConsent', None)
```
`Informed Consent` as this will be displayed for the end-user.

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/views/patients_listing.py:
```diff
@@ -89,6 +91,7 @@ def get_columns(self):
             ColumnDiagnosisProgress(_("Diagnosis Entry Progress"), "patients.can_see_diagnosis_progress"),
             ColumnDiagnosisCurrency(_("Updated < 365 days"), "patients.can_see_diagnosis_currency"),
             ColumnGeneticDataMap(_("Genetic Data"), "patients.can_see_genetic_data_map"),
+            ColumnPatientStage(_("Stage"), "patients.can_see_data_modules"),
```
Can we add this only if there is at least one registry with stages feature on?

alintreznai replied:
It's possible. Need to look into it.

sztamas commented on rdrf/registry/patients/models.py:
```diff
@@ -1118,6 +1125,15 @@ def family(self):
+class PatientStage(models.Model):
+    name = models.CharField(max_length=200)
+    allowed_prev_stages = models.ManyToManyField('self', symmetrical=False, related_name='prev_stages')
+    allowed_next_stages = models.ManyToManyField('self', symmetrical=False, related_name='next_stages')
```
`related_name` is the inverse relationship so they should be called accordingly.

alintreznai replied:
Updated.

sztamas committed:
Stages: show stages column only if stages features is enabled.

Commit:
Fix issue with initial data execution of patient_stage

Commit:
Added PatientStage to admin pages

sztamas committed:
Stages fixes:

  - order stages by PK
  - don't maintain the other side of the previous and next stages (we
keep both sides in the stage itself, so it is at least confusing and
probably incorrect as well to have the other side as well)
  - the query to find out the possible stages we can change a stage to
was wrong. Try adding more than 1 prev stage to reproduce.
